temporal-sdk/src/main/java/io/temporal/failure/ApplicationFailure.java

```diff
@@ -274,4 +246,122 @@ private static String getMessage(String message, String type, boolean nonRetryab
   }
+
+  public static final class Builder {
+    private String message;
+    private String type;
+    private Values details;
+    private boolean nonRetryable;
+    private Throwable cause;
+    private Duration nextRetryDelay;
+    private ApplicationErrorCategory category;
+
+    private Builder() {}
+
+    private Builder(ApplicationFailure options) {
+      if (options == null) {
+        return;
+      }
+      this.message = options.getOriginalMessage();
+      this.type = options.type;
+      this.details = options.details;
+      this.nonRetryable = options.nonRetryable;
+      this.nextRetryDelay = options.nextRetryDelay;
+      this.category = options.category;
+    }
+
+    /**
+     * Sets the error type of this failure. This is used by {@link
+     * io.temporal.common.RetryOptions.Builder#setDoNotRetry(String...)} to determine if the
+     * exception is non retryable.
+     */
+    public Builder setType(String type) {
+      this.type = type;
+      return this;
+    }
+
+    /**
+     * Set the optional error message.
+     *
+     * <p>Default is "".
+     */
+    public Builder setMessage(String message) {
+      this.message = message;
+      return this;
+    }
+
+    /**
+     * Set the optional details of the failure.
+     *
+     * <p>Details are serialized using the same approach as arguments and results.
+     */
+    public Builder setDetails(Object... details) {
+      this.details = new EncodedValues(details);
+      return this;
+    }
+
+    /**
+     * Set the optional details of the failure.
+     *
+     * <p>Details are serialized using the same approach as arguments and results.
+     */
+    public Builder setDetails(Values details) {
+      this.details = details;
+      return this;
+    }
+
+    /**
+     * Set the non retryable flag on the failure.
+     *
+     * <p>It means that this exception is not going to be retried even if it is not included into
+     * retry policy doNotRetry list.
+     *
+     * <p>Default is false.
+     */
+    public Builder setNonRetryable(boolean nonRetryable) {
+      this.nonRetryable = nonRetryable;
+      return this;
+    }
+
+    /**
+     * Set the optional cause of the failure. Each element of the cause chain will be converted to
+     * {@link ApplicationFailure} for network transmission across network if it doesn't extend
+     * {@link TemporalFailure}.
+     */
+    public Builder setCause(Throwable cause) {
+      this.cause = cause;
+      return this;
+    }
+
+    /**
+     * Set the optional delay before the next retry attempt. Overrides the normal retry delay.
+     *
+     * <p>Default is null.
+     */
+    public Builder setNextRetryDelay(Duration nextRetryDelay) {
+      this.nextRetryDelay = nextRetryDelay;
+      return this;
+    }
+
+    /**
+     * Set the optional category of the failure.
+     *
+     * <p>Default is {@link ApplicationErrorCategory#UNSPECIFIED}.
+     */
+    public Builder setCategory(ApplicationErrorCategory category) {
```

Comment:
I think we should change the getter on `ApplicationFailure` to `getCategory`

Reply:
I would also be fine changing the setter instead, any preference?

Reply:
+1 to the shorter version